### PR TITLE
A minimal Boost build 

### PR DESCRIPTION
**Pull request by dkeeney**
A boost download and build that executes at configure time and builds with -fPIC

### CODE REVIEW

File: external/Boost.cmake [hunk 1]
```diff
@@ -18,49 +18,112 @@
...
+
+
+
+
+# Download the boost distribution (at configure time).
+message(STATUS "obtaining Boost")
+include(DownloadProject/DownloadProject.cmake)
+
+  if (MSVC OR MSYS OR MINGW)
+    set(BOOST_URL "${REPOSITORY_DIR}/external/common/share/boost/boost_1_69_0_subset.zip")
```
Comment (breznak) [inline external/Boost.cmake]: so this will be how to enable off-line configuration :+1:

**Comment by breznak** [comment]
> What we don't know is if the resulting extension libraries will load in Python. For that we have to go back to the pybind11 PR. 

Ok, so I won't merge yet. And we test locally if this fixes #157

**Comment by breznak** [comment]
@dkeeney I found 2 problems
- [ ] cmake caches the download (at least), would be nice if could(?) cache also the boost b2 build in repetitive compilations
- [ ] calling `cmake` twice causes an error: 
> cmake $NUPIC_CORE -DCMAKE_BUILD_TYPE=Release 
// make install ..ok
cmake $NUPIC_CORE -DCMAKE_BUILD_TYPE=Release 
// 2nd time fails

`make clean` does not help, have to git clean -xdf and that forces the download again 

```

-- obtaining Boost ...may take a while
CMake Error at external/Boost.cmake:71 (list):
  list sub-command LENGTH requires two arguments.
Call Stack (most recent call first):
  external/CMakeLists.txt:39 (include)

```

**Comment by breznak** [comment]
> make caches the download (at least), would be nice if could(?) cache also the boost b2 build in repetitive compilations 

I think both problems could be solved if the Boost.cmake first checked for a boost installation (with fPIC) and only if not found attempted to configure & install Boost.

**Comment by breznak** [comment]
FYI

**Comment by dkeeney** [comment]
>  cmake caches the download (at least), would be nice if could(?) cache also the boost b2 build in repetitive compilations

Aware of that.  I thought I would do that for all dependancies in later PR.  Will have to put them in a ExpernalProject in order to get a separate cache.

**Comment by dkeeney** [comment]
> Problem is that setup.py calls the 2nd installation of Boost, and fails.
Would this work #159 (comment) , rather than the ext project?

Ok, let me try this some more.  Though I had it working.

**Comment by dkeeney** [comment]
Do you still want me to take the time to do the dependencies with a separate cache?

**Comment by breznak** [comment]
> Do you still want me to take the time to do the dependencies with a separate cache?

I think we can proceed w/o it, if the new fix allows pybind to build :)

**Comment by dkeeney** [comment]
hmmm,
`ld: warning: direct access in function 'boost::filesystem::detail::canonical(boost::filesystem::path const&, boost::filesystem::path const&, boost::system::error_code*)' from file 'libnupic_core.a(libboost_filesystem.a-operations.o)' to global weak symbol 'boost::system::generic_category()::generic_category_instance' from file 'CMakeFiles/benchmark_hotgym.dir/examples/Hotgym.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.`

Saw this warning in Circleci.  Do you think this might be a problem?  
Perhaps I need to set the visibility setting on the boost build?

**Comment by breznak** [comment]
> Perhaps I need to set the visibility setting on the boost build? 

We have visibility off, so maybe set to hidden. But let's see now. 
I think boost should compile with CXX flags and compiler settings from our main CMake

Mine linux errs at:
```
[ 58%] Building CXX object bindings/py/cpp_src/CMakeFiles/math.dir/bindings/math/py_Domain.cpp.o
In file included from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/math/ArrayAlgo.hpp:51:0,
                 from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/math/SparseMatrix.hpp:38,
                 from /home/mmm/devel/HTM/htm-community/nupic.cpp/bindings/py/cpp_src/bindings/math/py_ArrayAlgo.cpp:45:
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/math/Math.hpp:55:11: fatal error: boost/concept_check.hpp: Path or file does not exist
  #include <boost/concept_check.hpp>
```

**Comment by dkeeney** [comment]
I am only setting the one flag -fPIC,  Perhaps I should put them all in there.

**Pull request merged**